src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,167 @@
+    /// |
+    /// root: 32.000bytes
+    /// | |
+    /// | child1: 24.000bytes
+    /// | | |
+    /// | | child11: 0.000bytes
+    /// | | | |
+    /// | | | child111: 0.000bytes
+    /// | |
+    /// | child2: 0.000bytes
+    ss::sstring debug(int level = 0);
+
+private:
+    ss::sstring _label;
+    int64_t _consumption = 0;
+    std::list<ss::lw_shared_ptr<mem_tracker>> _children;
+};
+
+/// An allocator paired with a memory tracker to account for consumption.
+template<class T, class allocator = std::allocator<T>>
+class tracking_allocator : public allocator {
+public:
+    explicit tracking_allocator(ss::lw_shared_ptr<mem_tracker> tracker)
+      : _tracker(std::move(tracker)) {}
+
+    template<class U>
+    explicit tracking_allocator(const tracking_allocator<U>& other)
+      : allocator(other)
+      , _tracker(other.tracker()) {}
+
+    template<class U>
+    struct rebind {
+        using other = tracking_allocator<U, std::allocator<U>>;
+    };
+
+    [[nodiscard]] T* allocate(std::size_t n) {
+        _tracker->allocate(n * sizeof(T));
+        return allocator::allocate(n);
+    }
+
+    void deallocate(T* p, std::size_t n) {
+        allocator::deallocate(p, n);
+        _tracker->deallocate(n * sizeof(T));
+    }
+
+    ss::lw_shared_ptr<mem_tracker> tracker() const { return _tracker; }
+
+private:
+    ss::lw_shared_ptr<mem_tracker> _tracker;
+};
+
+/// Since allocator is typically the last parameter for stl containers,
+/// these template utilities help avoid template boiler plate by auto
+/// setting the intermediate parameters to defaults.
+template<class T>
+concept is_map = requires {
+    typename T::key_type;
+    typename T::value_type;
+    typename T::mapped_type;
+    typename T::hasher;
+    typename T::key_equal;
+    typename T::allocator_type;
+};
+
+template<class T>
+concept is_set = requires {
+    typename T::key_type;
+    typename T::value_type;
+    typename T::key_compare;
+    typename T::value_compare;
+    typename T::allocator_type;
+};
+
+template<
+  template<class...>
+  class Map,
+  class K,
+  class V,
+  class Hasher = typename Map<K, V>::hasher,
+  class KeyEq = typename Map<K, V>::key_equal,
+  class Allocator = tracking_allocator<std::pair<const K, V>>>
+using map_t = Map<K, V, Hasher, KeyEq, Allocator>;
+
+template<
+  template<class...>
+  class Set,
+  class K,
+  class Compare = std::less<K>,
+  class Allocator = tracking_allocator<K>>
+using set_t = Set<K, Compare, Allocator>;
+
+template<
+  template<class...>
+  class Map,
+  class K,
+  class V,
+  class Hasher = typename Map<K, V>::hasher,
+  class KeyEq = typename Map<K, V>::key_equal,
+  class Allocator = tracking_allocator<std::pair<const K, V>>>
+requires is_map<map_t<Map, K, V, Hasher, KeyEq, Allocator>>
+auto map(ss::lw_shared_ptr<mem_tracker> tracker)
+  -> map_t<Map, K, V, Hasher, KeyEq, Allocator> {
+    return Map<K, V, Hasher, KeyEq, Allocator>(
+      tracking_allocator<std::pair<const K, V>>{tracker});
+}
```

Comment:
`map` seems a bit too generic to be at the `util::map` level. should this (and the concepts too) be moved into a mem_tracker namespace or be given a less generic name at the util level?